transform-plugins/src/main/java/co/cask/hydrator/plugin/ValueMapper.java

```diff
@@ -0,0 +1,251 @@
+          valueMapping.setDefaultValue(defaultsMapping.get(mappingValueArray[0]));
+        }
+        mappingValues.put(mappingValueArray[0], valueMapping);
+      }
+    }
+  }
+
+  /**
+   * Creates output schema with the help of input schema and mapping
+   */
+  private Schema getOutputSchema(Schema inputSchema) throws IllegalArgumentException {
+    Schema outputSchema = schemaCache.get(inputSchema);
+    if (outputSchema != null) {
+      return outputSchema;
+    }
+
+    List<Schema.Field> outputFields = new ArrayList<>();
+    for (Schema.Field inputField : inputSchema.getFields()) {
+      if (mappingValues.containsKey(inputField.getName())) {
+        if (inputField.getSchema().getType() != Schema.Type.STRING && !inputField.getSchema().isNullable()) {
+          throw new IllegalArgumentException("Input field " + inputField.getName() + " type should be String");
+        } else {
+          outputFields.add(Schema.Field.of(mappingValues.get(inputField.getName()).getTargetField(),
+                                           inputField.getSchema()));
+          }
+      } else {
+        outputFields.add(inputField);
+      }
+    }
+    outputSchema = Schema.recordOf(inputSchema.getRecordName() + ".formatted", outputFields);
+    schemaCache.put(inputSchema, outputSchema);
+    return outputSchema;
+  }
+
+  /**
+   * retrieve lookup table from table name
+   */
+  private void setLookup(TransformContext context) {
+    for (String key :mappingValues.keySet()) {
+      ValueMapping mapping = mappingValues.get(key);
+      LookupTableConfig tableConfig = new LookupTableConfig(LookupTableConfig.TableType.DATASET);
+      DatasetProperties arguments = DatasetProperties.builder().addAll(tableConfig.getDatasetProperties()).build();
+      Lookup<String> lookupTable = context.provide(mapping.getLookupTableName(), arguments.getProperties());
+      mapping.setLookupTable(lookupTable);
+    }
+  }
+
+  @Override
+  public void transform(StructuredRecord input, Emitter<StructuredRecord> emitter) throws Exception {
+
+    StructuredRecord.Builder builder = StructuredRecord.builder(getOutputSchema(input.getSchema()));
+
+    for (Schema.Field sourceField : input.getSchema().getFields()) {
+      if (mappingValues.containsKey(sourceField.getName())) {
+        ValueMapping mapping = mappingValues.get(sourceField.getName());
+        if (input.get(sourceField.getName()) == null || input.get(sourceField.getName()).toString().isEmpty()) {
+          if (mapping.getDefaultValue() != null) {
+            builder.set(mapping.getTargetField(), mapping.getDefaultValue());
+          } else {
+            builder.set(mapping.getTargetField(), input.get(sourceField.getName()));
+          }
+        } else {
+          // for those source field whose values are neither NULL nor EMPTY
+          Lookup<String> valueMapperLookUp = mapping.getLookupTable();
+          String lookupValue = valueMapperLookUp.lookup(input.get(sourceField.getName()).toString());
```

Comment:
same here, should not call `.toString()`.  It is actually wrong here, can result in a NPE if the lookup value is null.